In [0]:
from pyspark.sql.functions import col, avg, count, when
import pandas as pd
import numpy as np
 
df_drivers   = spark.read.csv("/Volumes/gr5069/raw/f1_data/drivers.csv",   header=True, inferSchema=True)
df_races     = spark.read.csv("/Volumes/gr5069/raw/f1_data/races.csv",     header=True, inferSchema=True)
df_pit_stops = spark.read.csv("/Volumes/gr5069/raw/f1_data/pit_stops.csv", header=True, inferSchema=True)
df_results   = spark.read.csv("/Volumes/gr5069/raw/f1_data/results.csv",   header=True, inferSchema=True)
 
print("Done")

Done


In [0]:
pit_agg = df_pit_stops.groupBy("raceId", "driverId").agg(
    count("stop").alias("pit_stop_count"),
    avg(col("milliseconds").cast("double")).alias("avg_pit_ms"),
)
 
results_clean = (
    df_results
    .select(
        "raceId", "driverId", "constructorId",
        col("positionOrder").cast("int").alias("finish_pos"),
        col("grid").cast("int").alias("grid"),
        col("laps").cast("int").alias("laps"),
        col("statusId").cast("int").alias("statusId"),
    )
    .filter(col("finish_pos").isNotNull())
)
 
races_clean = df_races.select(
    "raceId",
    col("year").cast("int").alias("year"),
    col("round").cast("int").alias("round"),
)
 
df = (
    results_clean
    .join(races_clean, on="raceId", how="left")
    .join(pit_agg, on=["raceId", "driverId"], how="left")
    .join(df_drivers.select("driverId", "nationality"), on="driverId", how="left")
    .withColumn("points_finish", when(col("finish_pos") <= 10, 1).otherwise(0))
    .withColumn("front_row",     when(col("grid") <= 2, 1).otherwise(0))
    .withColumn("finished_race", when(col("statusId") == 1, 1).otherwise(0))
    .fillna(0)
)
 
print(f"Rows: {df.count():,}")

Rows: 26,759


In [0]:
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
 
pdf = df.toPandas()
pdf["nat_enc"] = LabelEncoder().fit_transform(pdf["nationality"].fillna("Unknown"))
 
FEATURES = [
    "grid", "front_row", "laps",
    "pit_stop_count", "avg_pit_ms",
    "finished_race", "year", "round",
    "constructorId", "nat_enc",
]
TARGET = "points_finish"
 
X = pdf[FEATURES].fillna(0)
y = pdf[TARGET]
 
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42, stratify=y
)
 
print(f"Train: {len(X_train):,} | Test: {len(X_test):,}")

Train: 21,407 | Test: 5,352


In [0]:
import mlflow
import mlflow.sklearn
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import cross_val_score
from sklearn.metrics import accuracy_score
from sklearn.metrics import precision_score
from sklearn.metrics import recall_score
from sklearn.metrics import f1_score
from sklearn.metrics import roc_auc_score
from sklearn.metrics import log_loss
from sklearn.metrics import matthews_corrcoef
from sklearn.metrics import balanced_accuracy_score
from sklearn.metrics import confusion_matrix
from mlflow.models.signature import infer_signature
 
mlflow.set_experiment("/Users/kc3949@columbia.edu/f1_model")
 
PARAM_GRID = [
    {"n_estimators": 50,  "max_depth": 3,    "min_samples_split": 2,  "class_weight": None},
    {"n_estimators": 100, "max_depth": 5,    "min_samples_split": 2,  "class_weight": None},
    {"n_estimators": 100, "max_depth": 10,   "min_samples_split": 5,  "class_weight": None},
    {"n_estimators": 150, "max_depth": 8,    "min_samples_split": 5,  "class_weight": None},
    {"n_estimators": 200, "max_depth": 12,   "min_samples_split": 10, "class_weight": None},
    {"n_estimators": 200, "max_depth": None, "min_samples_split": 2,  "class_weight": "balanced"},
    {"n_estimators": 100, "max_depth": 6,    "min_samples_split": 4,  "class_weight": "balanced"},
    {"n_estimators": 300, "max_depth": 10,   "min_samples_split": 5,  "class_weight": "balanced"},
    {"n_estimators": 150, "max_depth": 7,    "min_samples_split": 3,  "class_weight": None},
    {"n_estimators": 250, "max_depth": 15,   "min_samples_split": 8,  "class_weight": "balanced"},
    {"n_estimators": 500, "max_depth": 20,   "min_samples_split": 2,  "class_weight": None},
    {"n_estimators": 100, "max_depth": 4,    "min_samples_split": 6,  "class_weight": "balanced"},
]
 
results = []
 
for i, p in enumerate(PARAM_GRID, 1):
    run_name = "run_{:02d}_n{}_d{}".format(i, p["n_estimators"], p["max_depth"])
    with mlflow.start_run(run_name=run_name) as run:
        rid = run.info.run_id
 
        model = RandomForestClassifier(
            n_estimators=p["n_estimators"],
            max_depth=p["max_depth"],
            min_samples_split=p["min_samples_split"],
            min_samples_leaf=1,
            max_features="sqrt",
            class_weight=p["class_weight"],
            oob_score=True,
            random_state=42,
            n_jobs=-1,
        )
        model.fit(X_train, y_train)
 
        y_pred = model.predict(X_test)
        y_prob = model.predict_proba(X_test)[:, 1]
        cv_f1 = cross_val_score(model, X_train, y_train, cv=5, scoring="f1", n_jobs=-1)
 
        metrics = {
            "accuracy":          accuracy_score(y_test, y_pred),
            "precision":         precision_score(y_test, y_pred, zero_division=0),
            "recall":            recall_score(y_test, y_pred, zero_division=0),
            "f1":                f1_score(y_test, y_pred, zero_division=0),
            "roc_auc":           roc_auc_score(y_test, y_prob),
            "log_loss":          log_loss(y_test, y_prob),
            "mcc":               matthews_corrcoef(y_test, y_pred),
            "balanced_accuracy": balanced_accuracy_score(y_test, y_pred),
            "cv_f1_mean":        float(cv_f1.mean()),
            "cv_f1_std":         float(cv_f1.std()),
            "oob_score":         float(model.oob_score_),
        }
 
        mlflow.log_params(p)
        mlflow.log_metrics(metrics)
        mlflow.sklearn.log_model(
            model, "model",
            signature=infer_signature(X_train, y_pred),
            input_example=X_train.head(3),
        )
 
        fig, ax = plt.subplots(figsize=(4, 3))
        sns.heatmap(
            confusion_matrix(y_test, y_pred),
            annot=True, fmt="d",
            xticklabels=["No", "Yes"],
            yticklabels=["No", "Yes"],
            ax=ax,
        )
        ax.set_title("Confusion Matrix")
        ax.set_xlabel("Predicted")
        ax.set_ylabel("Actual")
        fig.tight_layout()
        fig.savefig("/tmp/cm_{}.png".format(rid[:6]))
        plt.close()
        mlflow.log_artifact("/tmp/cm_{}.png".format(rid[:6]), "plots")
 
        imp = pd.Series(model.feature_importances_, index=FEATURES).sort_values()
        fig, ax = plt.subplots(figsize=(6, 4))
        imp.plot(kind="barh", ax=ax, color="steelblue")
        ax.set_title("Feature Importances")
        fig.tight_layout()
        fig.savefig("/tmp/fi_{}.png".format(rid[:6]))
        plt.close()
        mlflow.log_artifact("/tmp/fi_{}.png".format(rid[:6]), "plots")
 
        pd.DataFrame([metrics]).to_csv("/tmp/metrics_{}.csv".format(rid[:6]), index=False)
        mlflow.log_artifact("/tmp/metrics_{}.csv".format(rid[:6]), "csv")
 
        pd.DataFrame({
            "actual": y_test.values,
            "predicted": y_pred,
            "prob": y_prob,
        }).to_csv("/tmp/preds_{}.csv".format(rid[:6]), index=False)
        mlflow.log_artifact("/tmp/preds_{}.csv".format(rid[:6]), "csv")
 
        results.append({"run": "run_{:02d}".format(i), **metrics, **p})
        print("run_{:02d} | AUC={:.3f} | F1={:.3f} | Acc={:.3f}".format(
            i, metrics["roc_auc"], metrics["f1"], metrics["accuracy"]
        ))
 
print("All done!")

/databricks/python/lib/python3.12/site-packages/mlflow/types/utils.py:452: UserWarning: Hint: Inferred schema contains integer column(s). Integer columns in Python cannot represent missing values. If your input data contains missing values at inference time, it will be encoded as floats and will cause a schema enforcement error. The best way to avoid this problem is to infer the model schema based on a realistic data sample (training dataset) that includes missing values. Alternatively, you can declare integer columns as doubles (float64) whenever these columns may have missing values. See `Handling Integers With Missing Values <https://www.mlflow.org/docs/latest/models.html#handling-integers-with-missing-values>`_ for more details.
  warnings.warn(
2026/04/14 06:58:33 WARNING mlflow.models.model: `artifact_path` is deprecated. Please use `name` instead.
🔗 View Logged Model at: https://dbc-f11e6b56-7522.cloud.databricks.com/ml/experiments/4359974419176185/models/m-df357ba6dd96489cac696

run_01 | AUC=0.925 | F1=0.787 | Acc=0.833


/databricks/python/lib/python3.12/site-packages/mlflow/types/utils.py:452: UserWarning: Hint: Inferred schema contains integer column(s). Integer columns in Python cannot represent missing values. If your input data contains missing values at inference time, it will be encoded as floats and will cause a schema enforcement error. The best way to avoid this problem is to infer the model schema based on a realistic data sample (training dataset) that includes missing values. Alternatively, you can declare integer columns as doubles (float64) whenever these columns may have missing values. See `Handling Integers With Missing Values <https://www.mlflow.org/docs/latest/models.html#handling-integers-with-missing-values>`_ for more details.
  warnings.warn(
2026/04/14 06:58:48 WARNING mlflow.models.model: `artifact_path` is deprecated. Please use `name` instead.
🔗 View Logged Model at: https://dbc-f11e6b56-7522.cloud.databricks.com/ml/experiments/4359974419176185/models/m-a8142f3820fb40ce8a73b

run_02 | AUC=0.937 | F1=0.822 | Acc=0.855


/databricks/python/lib/python3.12/site-packages/mlflow/types/utils.py:452: UserWarning: Hint: Inferred schema contains integer column(s). Integer columns in Python cannot represent missing values. If your input data contains missing values at inference time, it will be encoded as floats and will cause a schema enforcement error. The best way to avoid this problem is to infer the model schema based on a realistic data sample (training dataset) that includes missing values. Alternatively, you can declare integer columns as doubles (float64) whenever these columns may have missing values. See `Handling Integers With Missing Values <https://www.mlflow.org/docs/latest/models.html#handling-integers-with-missing-values>`_ for more details.
  warnings.warn(
2026/04/14 06:59:01 WARNING mlflow.models.model: `artifact_path` is deprecated. Please use `name` instead.
🔗 View Logged Model at: https://dbc-f11e6b56-7522.cloud.databricks.com/ml/experiments/4359974419176185/models/m-6c7cb1b58ec44060a1c44

run_03 | AUC=0.950 | F1=0.841 | Acc=0.869


/databricks/python/lib/python3.12/site-packages/mlflow/types/utils.py:452: UserWarning: Hint: Inferred schema contains integer column(s). Integer columns in Python cannot represent missing values. If your input data contains missing values at inference time, it will be encoded as floats and will cause a schema enforcement error. The best way to avoid this problem is to infer the model schema based on a realistic data sample (training dataset) that includes missing values. Alternatively, you can declare integer columns as doubles (float64) whenever these columns may have missing values. See `Handling Integers With Missing Values <https://www.mlflow.org/docs/latest/models.html#handling-integers-with-missing-values>`_ for more details.
  warnings.warn(
2026/04/14 06:59:20 WARNING mlflow.models.model: `artifact_path` is deprecated. Please use `name` instead.
🔗 View Logged Model at: https://dbc-f11e6b56-7522.cloud.databricks.com/ml/experiments/4359974419176185/models/m-9d8ebdc1ae794e409a384

run_04 | AUC=0.946 | F1=0.834 | Acc=0.864


/databricks/python/lib/python3.12/site-packages/mlflow/types/utils.py:452: UserWarning: Hint: Inferred schema contains integer column(s). Integer columns in Python cannot represent missing values. If your input data contains missing values at inference time, it will be encoded as floats and will cause a schema enforcement error. The best way to avoid this problem is to infer the model schema based on a realistic data sample (training dataset) that includes missing values. Alternatively, you can declare integer columns as doubles (float64) whenever these columns may have missing values. See `Handling Integers With Missing Values <https://www.mlflow.org/docs/latest/models.html#handling-integers-with-missing-values>`_ for more details.
  warnings.warn(
2026/04/14 06:59:41 WARNING mlflow.models.model: `artifact_path` is deprecated. Please use `name` instead.
🔗 View Logged Model at: https://dbc-f11e6b56-7522.cloud.databricks.com/ml/experiments/4359974419176185/models/m-21cf1cb12bca496aa6399

run_05 | AUC=0.952 | F1=0.846 | Acc=0.872


/databricks/python/lib/python3.12/site-packages/mlflow/types/utils.py:452: UserWarning: Hint: Inferred schema contains integer column(s). Integer columns in Python cannot represent missing values. If your input data contains missing values at inference time, it will be encoded as floats and will cause a schema enforcement error. The best way to avoid this problem is to infer the model schema based on a realistic data sample (training dataset) that includes missing values. Alternatively, you can declare integer columns as doubles (float64) whenever these columns may have missing values. See `Handling Integers With Missing Values <https://www.mlflow.org/docs/latest/models.html#handling-integers-with-missing-values>`_ for more details.
  warnings.warn(
2026/04/14 07:00:07 WARNING mlflow.models.model: `artifact_path` is deprecated. Please use `name` instead.
🔗 View Logged Model at: https://dbc-f11e6b56-7522.cloud.databricks.com/ml/experiments/4359974419176185/models/m-312ff62566644aefa4b71

run_06 | AUC=0.955 | F1=0.855 | Acc=0.879


/databricks/python/lib/python3.12/site-packages/mlflow/types/utils.py:452: UserWarning: Hint: Inferred schema contains integer column(s). Integer columns in Python cannot represent missing values. If your input data contains missing values at inference time, it will be encoded as floats and will cause a schema enforcement error. The best way to avoid this problem is to infer the model schema based on a realistic data sample (training dataset) that includes missing values. Alternatively, you can declare integer columns as doubles (float64) whenever these columns may have missing values. See `Handling Integers With Missing Values <https://www.mlflow.org/docs/latest/models.html#handling-integers-with-missing-values>`_ for more details.
  warnings.warn(
2026/04/14 07:00:24 WARNING mlflow.models.model: `artifact_path` is deprecated. Please use `name` instead.
🔗 View Logged Model at: https://dbc-f11e6b56-7522.cloud.databricks.com/ml/experiments/4359974419176185/models/m-8e986429459c418ca18e2

run_07 | AUC=0.940 | F1=0.833 | Acc=0.854


/databricks/python/lib/python3.12/site-packages/mlflow/types/utils.py:452: UserWarning: Hint: Inferred schema contains integer column(s). Integer columns in Python cannot represent missing values. If your input data contains missing values at inference time, it will be encoded as floats and will cause a schema enforcement error. The best way to avoid this problem is to infer the model schema based on a realistic data sample (training dataset) that includes missing values. Alternatively, you can declare integer columns as doubles (float64) whenever these columns may have missing values. See `Handling Integers With Missing Values <https://www.mlflow.org/docs/latest/models.html#handling-integers-with-missing-values>`_ for more details.
  warnings.warn(
2026/04/14 07:00:51 WARNING mlflow.models.model: `artifact_path` is deprecated. Please use `name` instead.
🔗 View Logged Model at: https://dbc-f11e6b56-7522.cloud.databricks.com/ml/experiments/4359974419176185/models/m-fb776b62b8ae4a57b7062

run_08 | AUC=0.950 | F1=0.849 | Acc=0.870


/databricks/python/lib/python3.12/site-packages/mlflow/types/utils.py:452: UserWarning: Hint: Inferred schema contains integer column(s). Integer columns in Python cannot represent missing values. If your input data contains missing values at inference time, it will be encoded as floats and will cause a schema enforcement error. The best way to avoid this problem is to infer the model schema based on a realistic data sample (training dataset) that includes missing values. Alternatively, you can declare integer columns as doubles (float64) whenever these columns may have missing values. See `Handling Integers With Missing Values <https://www.mlflow.org/docs/latest/models.html#handling-integers-with-missing-values>`_ for more details.
  warnings.warn(
2026/04/14 07:01:09 WARNING mlflow.models.model: `artifact_path` is deprecated. Please use `name` instead.
🔗 View Logged Model at: https://dbc-f11e6b56-7522.cloud.databricks.com/ml/experiments/4359974419176185/models/m-5f5bf6f2f22a424b96d37

run_09 | AUC=0.943 | F1=0.828 | Acc=0.860


/databricks/python/lib/python3.12/site-packages/mlflow/types/utils.py:452: UserWarning: Hint: Inferred schema contains integer column(s). Integer columns in Python cannot represent missing values. If your input data contains missing values at inference time, it will be encoded as floats and will cause a schema enforcement error. The best way to avoid this problem is to infer the model schema based on a realistic data sample (training dataset) that includes missing values. Alternatively, you can declare integer columns as doubles (float64) whenever these columns may have missing values. See `Handling Integers With Missing Values <https://www.mlflow.org/docs/latest/models.html#handling-integers-with-missing-values>`_ for more details.
  warnings.warn(
2026/04/14 07:01:36 WARNING mlflow.models.model: `artifact_path` is deprecated. Please use `name` instead.
🔗 View Logged Model at: https://dbc-f11e6b56-7522.cloud.databricks.com/ml/experiments/4359974419176185/models/m-e58c74cd495649349c500

run_10 | AUC=0.955 | F1=0.855 | Acc=0.876


/databricks/python/lib/python3.12/site-packages/joblib/externals/loky/process_executor.py:752: UserWarning: A worker stopped while some jobs were given to the executor. This can be caused by a too short worker timeout or by a memory leak.
  warnings.warn(
/databricks/python/lib/python3.12/site-packages/mlflow/types/utils.py:452: UserWarning: Hint: Inferred schema contains integer column(s). Integer columns in Python cannot represent missing values. If your input data contains missing values at inference time, it will be encoded as floats and will cause a schema enforcement error. The best way to avoid this problem is to infer the model schema based on a realistic data sample (training dataset) that includes missing values. Alternatively, you can declare integer columns as doubles (float64) whenever these columns may have missing values. See `Handling Integers With Missing Values <https://www.mlflow.org/docs/latest/models.html#handling-integers-with-missing-values>`_ for more details.
 

run_11 | AUC=0.955 | F1=0.855 | Acc=0.879


/databricks/python/lib/python3.12/site-packages/mlflow/types/utils.py:452: UserWarning: Hint: Inferred schema contains integer column(s). Integer columns in Python cannot represent missing values. If your input data contains missing values at inference time, it will be encoded as floats and will cause a schema enforcement error. The best way to avoid this problem is to infer the model schema based on a realistic data sample (training dataset) that includes missing values. Alternatively, you can declare integer columns as doubles (float64) whenever these columns may have missing values. See `Handling Integers With Missing Values <https://www.mlflow.org/docs/latest/models.html#handling-integers-with-missing-values>`_ for more details.
  warnings.warn(
2026/04/14 07:02:48 WARNING mlflow.models.model: `artifact_path` is deprecated. Please use `name` instead.
🔗 View Logged Model at: https://dbc-f11e6b56-7522.cloud.databricks.com/ml/experiments/4359974419176185/models/m-f6ce7d5a9a99464784521

run_12 | AUC=0.933 | F1=0.824 | Acc=0.842
All done!


In [0]:
summary = pd.DataFrame(results).sort_values("roc_auc", ascending=False).reset_index(drop=True)
 
print(summary[["run", "roc_auc", "f1", "accuracy", "mcc", "n_estimators", "max_depth", "class_weight"]].to_string(index=False))
 
best = summary.iloc[0]
print("\nBest run: {}".format(best["run"]))
print("  ROC-AUC : {:.4f}".format(best["roc_auc"]))
print("  F1      : {:.4f}".format(best["f1"]))
print("  Accuracy: {:.4f}".format(best["accuracy"]))
print("  MCC     : {:.4f}".format(best["mcc"]))
print("  n_estimators={}, max_depth={}, class_weight={}".format(
    best["n_estimators"], best["max_depth"], best["class_weight"]
))
print("""
Why best?
- ROC-AUC is the primary metric: dataset is imbalanced (~30% points finishers)
  so accuracy alone is misleading. ROC-AUC measures discrimination at all thresholds.
- High F1 confirms balance between precision and recall.
- MCC > 0.5 means genuine signal beyond majority-class baseline.
- CV F1 close to hold-out F1 means the model generalises well.
""")
 

   run  roc_auc       f1  accuracy      mcc  n_estimators  max_depth class_weight
run_11 0.955303 0.855293  0.878924 0.751374           500       20.0         None
run_06 0.954737 0.854839  0.878924 0.751256           200        NaN     balanced
run_10 0.954559 0.855139  0.876495 0.747580           250       15.0     balanced
run_05 0.952395 0.845528  0.872197 0.737183           200       12.0         None
run_08 0.949988 0.849392  0.870329 0.735949           300       10.0     balanced
run_03 0.949935 0.840527  0.868834 0.730151           100       10.0         None
run_04 0.945634 0.834208  0.864163 0.720463           150        8.0         None
run_09 0.942970 0.828473  0.860426 0.712726           150        7.0         None
run_07 0.940172 0.833262  0.854260 0.705202           100        6.0     balanced
run_02 0.936715 0.822400  0.854821 0.701111           100        5.0         None
run_12 0.932751 0.824408  0.841928 0.685718           100        4.0     balanced
run_01 0.924894 

Best Model: run_11
Hyperparameters: n_estimators=500, max_depth=20, class_weight=None
Metrics:

ROC-AUC: 0.9553 

 # primary metric
F1: 0.8553
Accuracy: 0.8789
MCC: 0.7514

Why run_11 is the best?

1. Highest ROC-AUC (0.9553)
ROC-AUC is the primary selection metric because the dataset is class-imbalanced - only ~30% of race entries finish in the points (top 10). Accuracy alone is misleading; a model that always predicts "no points" would still score ~70%. ROC-AUC measures the model's ability to correctly rank points finishers above non-finishers across every possible threshold, making it the most reliable metric here.
2. Best F1 Score (0.8553)
F1 is the harmonic mean of precision and recall. run_11 achieves the highest F1, meaning it best balances correctly identifying points finishers (recall) without too many false alarms (precision).
3. Highest MCC (0.7514)
Matthews Correlation Coefficient above 0.5 indicates genuine predictive signal well beyond a naive majority-class baseline. It is symmetric and robust to class imbalance, providing a second independent confirmation.
4. Why n_estimators=500, max_depth=20 works best
A larger ensemble (500 trees) reduces variance through averaging — individual trees may overfit, but the ensemble corrects this. The deeper max_depth=20 allows each tree to capture complex non-linear patterns in the data (e.g., the interaction between grid position, pit stop strategy, and circuit), which a shallow tree would miss. Together they achieve the best bias-variance tradeoff in this experiment.Sonnet 4.6